### PR TITLE
[desktop] add manual update controls to desktop settings

### LINKED ISSUE #2151: Add more information about updates to desktop settings
- show if there's an update ready to be installed
- add "check for update now" button

### PR DESCRIPTION
**Pull request by ganthern**
the help text for the auto updater setting now includes a button to
check for an update for the desktop client / install the update if it
is already available

this label will be automatically updated if the updater downloads an
update while the settings menu is opened

autoupdaters

close #2151

also contains better names for desktop config access methods in 6939217

### CODE REVIEW

File: src/desktop/ElectronUpdater.js [hunk 1]
```diff
@@ -168,47 +184,101 @@ export class ElectronUpdater {
...
-				    app.emit('enable-force-quit')
-				    this._installOnQuit = false
-				    autoUpdater.quitAndInstall(false, true)
+				    this.installUpdate()
 			    }
 		    })
 		    .catch((e: Error) => this._logger.error("Notification failed,", e.message))
 	}
 
+	/**
```
Comment (charlag) [inline src/desktop/ElectronUpdater.js:262]: Should we solve this before merging or do you want to do it separately?

File: src/desktop/ElectronUpdater.js [hunk 2]
```diff
@@ -168,47 +184,101 @@ export class ElectronUpdater {
...
+			autoUpdater.once('update-not-available', updateNotAvailable)
+			           .once('update-downloaded', updateDownloaded)
+			           .once('error', updateNotAvailable)
+		})
+	}
+
+	/**
+	 * check for update if none is currently available,
+	 * quit and install otherwise
+	 * @returns {Promise<boolean>}
```
Comment (charlag) [inline src/desktop/ElectronUpdater.js:224]: It's not clear what it returns

File: src/settings/DesktopUpdateHelpLabel.js [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+import m from "mithril"
+import {Icon} from "../gui/base/Icon"
+import {noOp} from "../api/common/utils/Utils"
+import {lang} from "../misc/LanguageViewModel"
+import {BootIcons} from "../gui/base/icons/BootIcons"
+import {nativeApp} from '../native/NativeWrapper.js'
+import {Request} from "../api/common/WorkerProtocol.js"
+
+export type UpdateHelpLabelAttrs = {
+	updateAvailable: Stream<boolean>;
```
Comment (charlag) [inline src/settings/DesktopUpdateHelpLabel.js:12]: We never subscribe to this stream inside, I think we should just pass boolean as an attr

File: src/settings/DesktopUpdateHelpLabel.js [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+	getIcon(): Child {
+		return this._waiting && !this._error
+			? m(Icon, {
+				icon: BootIcons.Progress,
+				class: 'flex-center items-center icon-progress-tiny icon-progress'
+			})
+			: null
+	}
+
+	getOnClick(updateAvailable: Stream<boolean>): () => void {
```
Comment (charlag) [inline src/settings/DesktopUpdateHelpLabel.js:54]: I think it's odd to decide which function to return, I would rather add `onClick()` method and decided what to do inside of it when the click actually happens.

File: src/settings/DesktopSettingsViewer.js [hunk 5]
```diff
@@ -226,6 +236,11 @@ export class DesktopSettingsViewer implements UpdatableSettingsViewer {
 		       })
 	}
 
+	appUpdate(): void {
```
Comment (charlag) [inline src/settings/DesktopSettingsViewer.js:239]: I would rename it into something like `onAppUpdateAvailable()`

File: src/settings/DesktopSettingsViewer.js [hunk 6]
```diff
@@ -33,6 +35,7 @@ export class DesktopSettingsViewer implements UpdatableSettingsViewer {
 	_isIntegrated: Stream<?boolean>;
 	_isAutoUpdateEnabled: Stream<?boolean>;
 	_showAutoUpdateOption: Stream<?boolean>;
```
Comment (charlag) [inline src/settings/DesktopSettingsViewer.js:37]: I think neither `_showAutoUpdateOption` nor `_updateAvailable` have to be streams

**Review by charlag — COMMENTED** [review]
I don't see an option. It must be a custom release? Maybe we should document this/how to test this.

File: src/settings/DesktopUpdateHelpLabel.js (hunk 3 above)
Comment (ganthern) [reply src/settings/DesktopUpdateHelpLabel.js:12]: it's changed from within getOnClick and is then used in the SettingsViewer for the next redraw 
https://github.com/tutao/tutanota/blob/696cff30692ce12b4f58c2a879639927fb73155b/src/settings/DesktopUpdateHelpLabel.js#L64-L76

the attrs type could be Stream<boolean>, but I don't see a way to return the state of the update check to the settings viewer as easily with just a bool.

File: src/settings/DesktopSettingsViewer.js (hunk 5 above)
Comment (ganthern) [reply src/settings/DesktopSettingsViewer.js:239]: agreed

File: src/settings/DesktopUpdateHelpLabel.js (hunk 3 above)
Comment (charlag) [reply src/settings/DesktopUpdateHelpLabel.js:12]: Ah, so it's for propagating back? my bad, I missed it
I think callback is more clear (and I would change more places where we use streams for callbacks in our code) but I don't have a strong opinion about this place.

File: src/desktop/ElectronUpdater.js (hunk 1 above)
Comment (charlag) [reply src/desktop/ElectronUpdater.js:262]: Okay, let's do it separately then

**Pull request merged**